dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,734 @@
+The order of dimensions is determined by the layout provided in ``shape_layout``.
 
-  // Copy the anchor to output
-  auto &anchor_out = ws.Output<CPUBackend>(0);
-  anchor_out.Resize({2});
+Note: ``crop_shape`` and ``input_shape`` should be provided together, and providing those is
+incompatible with using ``scaling`` and ``aspect_ratio`` arguments.)code",
+        std::vector<int>{}, true)
+    .AddOptionalArg<int>("input_shape",
+                         R"code(Specifies the shape of the original input image.
 
-  auto *anchor_out_data = anchor_out.mutable_data<float>();
-  anchor_out_data[0] = coordinates[0];
-  anchor_out_data[1] = coordinates[1];
+The order of dimensions is determined by the layout provided in ``shape_layout``.
 
-  // Copy the offsets to output 1
-  auto &offsets_out = ws.Output<CPUBackend>(1);
-  offsets_out.Resize({2});
+Note: ``crop_shape`` and ``input_shape`` should be provided together, and providing those is
+incompatible with using ``scaling`` and ``aspect_ratio`` arguments.)code",
+                         std::vector<int>{}, true)
+    .AddOptionalArg<TensorLayout>(
+        "bbox_layout",
+        R"code(Determines the meaning of the coordinates of the bounding boxes.
+  Possible values are:
 
-  auto *offsets_out_data = offsets_out.mutable_data<float>();
-  offsets_out_data[0] = coordinates[2];
-  offsets_out_data[1] = coordinates[3];
-}
+  ``x`` (horizontal start anchor), ``y`` (vertical start anchor), ``z`` (depthwise start anchor),
 
-template <>
-void RandomBBoxCrop<CPUBackend>::WriteBoxesToOutput(
-    SampleWorkspace &ws, const BoundingBoxes &bounding_boxes) const {
-  auto &bbox_out = ws.Output<CPUBackend>(2);
-  bbox_out.Resize(
-      {static_cast<int64_t>(bounding_boxes.size()), static_cast<int64_t>(BoundingBox::kSize)});
-
-  auto *bbox_out_data = bbox_out.mutable_data<float>();
-  for (size_t i = 0; i < bounding_boxes.size(); ++i) {
-    auto *output = bbox_out_data + i * BoundingBox::kSize;
-    auto coordinates =
-        ltrb_ ? bounding_boxes[i].AsLtrb() : bounding_boxes[i].AsXywh();
-    output[0] = coordinates[0];
-    output[1] = coordinates[1];
-    output[2] = coordinates[2];
-    output[3] = coordinates[3];
+  ``X`` (horizontal end anchor),   ``Y`` (vertical end anchor),   ``Z`` (depthwise end anchor),
+
+  ``W`` (width),                   ``H`` (height),                ``D`` (depth).
+
+Note: If left empty, ``"xyXY"`` or ``"xyzXYZ"`` will be assumed, depending on the number of dimensions.
+)code",
+        TensorLayout{""})
+    .AddOptionalArg<TensorLayout>(
+        "shape_layout",
+        R"code(Determines the meaning of the dimensions provided in `crop_shape` and `input_shape`.
+Possible values are:
+
+  ``W`` (width), ``H`` (height), ``D`` (depth).
+
+Note: If left empty, ``"WH"`` or ``"WHD"`` will be assumed, depending on the number of dimensions.
+)code",
+        TensorLayout{""});
+
+template <int ndim>
+class RandomBBoxCropImpl : public detail::OpImplBase<CPUBackend> {
```

Comment:
I couldn't find any tests for functions:`FixAspectRatios`, `FindProspectiveCrop`, `ValidAspectRatio`, `ValidOverlap`, `FilterByCentroid`. Are they written? I'd consider them necessary, as these functions are quite complex, and might have some corner-cases, which could be cought when we have the test and see, if they actually work as they should. Additionally, they could serve as an example and documentation of the function, which there are not too much also

Reply:
There are no such tests now. What we have are operator tests, both C++ and python, and they test quite extensively the output of the crops (they expect exact same result as previous SSDCrop operator). What I mean is that I checked carefully that the 2D case was not broken with this PR. I agree that those unit tests would be nice to have, but I feel that this is going a little bit far beyond the scope of the planned work for this task. The idea was to do *minimal* changes only to make the 3D bounding box crop case possible. We have already identified some major problems with the way we generate random crop that we should be handling in a follow up (it will introduce breaking changes), and I think that would be the time to add more extensive implementation detail testing

Reply:
Yeah, the problem is that changes turned out not to be that small ;) Fair enough, probably we should've spotted this at the beginning, that change won't be small, so it needs separate tasks...